### ISSUE TITLE
lack of documentation on how to refactor

### ISSUE THREAD

**Issue by benbarrowes** [camfort/camfort#152]
I can't seem to find any documentation on how to refactor a simple file. If I go to the wiki page:
https://github.com/camfort/camfort/wiki

It claims camfort *can* refactor, but no examples of how to do this from the command line in bash (for example) are provided anywhere I can find.

Poking around (using --help and error messages), I managed to try camfort on this file, test7.f90:

program test7

   common /abc/ var1(10)
   real var1
      real :: aa,dd,xx(6),yy(0:10),zz(3,5)
      real*8 rrr(3,3,3)
      !integer, parameter :: vlen=11,l2=2
      real, parameter :: v7=7.1
      integer ii,foo(11),i,bb(2,3)
      integer*8 surf(0:5), psd(3)
      character*64 cc,h3H,ee(2),suma(5)
      logical :: remg(1)=.true.
      logical lll(0:5),logi(2)
      integer jj(5)/10,20,30,40,30/
      character*6 ff

      print *,'aa=',aa
      cc='and who willbe me??--'
      lll(1)=.false.
      h3h='  '

      ii=12
      open(ii,file='input1')
      read(ii,'(A5)') cc,h3h
      print *,'cc=',cc
      print *,'h3h=',h3h
      close(ii)

      var1(2)=5
      print *,'var1=',var1
      
      end program test7

I inferred this command from poking around:
> camfort common --inplace test7.f90
Refactoring common blocks in 'test7.f90'
Successfully parsed 0 summary file(s).
test7.f90, 16:21: test7.f90: parsing failed. 
Last parsed token: TOpDivision (16:20)-(16:20).
Finished running common block refactoring on input '<unknown>' ...
Logs:

Result... OK:

I have no idea what this means or what I am doing wrong or what most of these CLI switches do.

Have I missed a nice pdf doc file somewhere?

**Comment by benbarrowes** [comment]
That syntax is not allowed in declarations, only in array definitions in the body. However, I simply commented out that line, and now this works:
camfort common --inplace test7.f90

Next, I'll try an equivalence example.

**Comment by benbarrowes** [comment]
Trying to test equivalent refactoring now. Here is a slightly modified test7 from above:

program test7

   use Abc, var1 => Abc_var1
      real :: aa,dd,xx(6),xx1(6),yy(0:10),zz(3,5)
      equivalence(xx,xx1)
      real*8 rrr(3,3,3)
      !integer, parameter :: vlen=11,l2=2
      real, parameter :: v7=7.1
      integer ii,foo(11),i,bb(2,3)
      integer*8 surf(0:5), psd(3)
      character*64 cc,h3H,ee(2),suma(5)
      logical :: remg(1)=.true.
      logical lll(0:5),logi(2)
      !integer jj(5) /10,20,30,40,30/
      character*6 ff

      print *,'aa=',aa
      cc='and who willbe me??--'
      lll(1)=.false.
      h3h='  '

      ii=12
      open(ii,file='input1')
      read(ii,'(A5)') cc,h3h
      print *,'cc=',cc
      print *,'h3h=',h3h
      close(ii)

      var1(2)=5
      print *,'var1=',var1
      xx(3)=3.1
      print *,'xx1=',xx1
      
      end program test7


And so I try camfort:

camfort equivalence --inplace test7.f90


But the result is the original code with the equivalence line simply removed. This breaks the code (incorrect results for xx1).

Am I doing something wrong.

**Comment by raehik** [comment]
No, you've used it correctly. I believe this is due to the equivalence elimination not properly inspecting all usages of the variable it eliminates. It doesn't detect that `xx(3)` is reassigning a value in `xx` (arrays aren't handled properly). If it did, it should insert `xx1 = xx` after `xx(3) = 3.1`. Apologies -- the equivalence elimination isn't quite complete.

---

For debugging, the most relevant code is here:

https://github.com/camfort/camfort/blob/15a69c0e4f5f6c88e65f48c0c5696dfa62eea2c8/src/Camfort/Transformation/EquivalenceElim.hs#L75-L83

`equivalentsToExpr` tests for `Expression` equality -- which means a subscript `xx(y)` won't match with a plain `xx`. Altering the check there should be the first step.

**Comment by benbarrowes** [comment]
I wrote a fortran to matlab converter (f2matlab). And I have tried to solve this equivalence issue as part of the conversion, but it is difficult. Eventually, I created a pointer class in matlab to take care of these issues, but it is not very matlab-ic. Generally, when someone gives me fortran code to convert, I have to buckle down and refactor the equivalence's by hand (hoping they are not too egregious).
